File: akka-samples/scala-samples/src/main/scala/actors/ask/AskFromOutside.scala
package actors.ask

import akka.actor.typed.scaladsl.AskPattern.{Askable, schedulerFromActorSystem}
import akka.actor.typed.scaladsl.Behaviors
import akka.actor.typed.{ActorRef, ActorSystem, Behavior}
import akka.util.Timeout
import actors.ask.GreetingActor.Greet

import scala.concurrent.duration.DurationInt
import scala.concurrent.{ExecutionContextExecutor, Future}
import scala.util.{Failure, Success}

object AskFromOutside {
  var system: ActorSystem[String] = _

  def main(args: Array[String]): Unit = {
    system = ActorSystem[String](guardian(), "ask-from-outside")
  }

  def ask(actorRef: ActorRef[Greet]): Unit = {
    implicit val timeout: Timeout = 3.seconds
    implicit val system: ActorSystem[String] = AskFromOutside.system
    implicit val ec: ExecutionContextExecutor = system.executionContext

    val f: Future[String] = actorRef.ask[String](replyTo => Greet("Danny", replyTo))
    f.onComplete {
      case Success(value) => system ! value
      case Failure(exception) => system ! s"Something went wrong! ($exception)"
    }
  }

  def guardian(): Behavior[String] =
    Behaviors.setup[String] { context =>
      ask(context.spawn(GreetingActor(), "dummy-greeter"))

      Behaviors.receive[String] {(context, message) =>
        println(message)
        Behaviors.stopped
      }
    }
}

File: akka-samples/java-samples/src/main/java/actors/ask/AskFromOutside.java
package actors.ask;

import akka.actor.typed.ActorRef;
import akka.actor.typed.ActorSystem;
import akka.actor.typed.Behavior;
import akka.actor.typed.javadsl.*;

import java.time.Duration;
import java.util.concurrent.CompletionStage;

public class AskFromOutside extends AbstractBehavior<String> {

    @Override
    public Receive<String> createReceive() {
        return newReceiveBuilder().onMessage(String.class, (message) -> {
            System.out.println(message);

            return Behaviors.stopped();
        }).build();
    }

    public AskFromOutside(ActorContext<String> context) {
        super(context);

        ask(context.getSystem(), context.spawn(GreetingActor.create(), "dummy-greeter"));
    }

    public void ask(ActorSystem<Void> system, ActorRef<GreetingActor.Greet> actorRef) {
        CompletionStage<String> result =
                AskPattern.ask(actorRef, (ActorRef<String> replyTo) ->
                        new GreetingActor.Greet("Danny", replyTo), Duration.ofSeconds(3), system.scheduler());

        result.whenComplete((reply, failure) -> {
           if (failure == null) {
               system.unsafeUpcast().tell(reply);
           } else {
               system.unsafeUpcast().tell(String.format("Something went wrong! (%s)", failure));
           }
        });
    }

    public static Behavior<String> create() {
        return Behaviors.setup(AskFromOutside::new);
    }

    public static void main(String[] args) {
        ActorSystem.create(AskFromOutside.create(), "ask-from-outside");
    }
}
